docs/developer/best-practices/security.asciidoc

```diff
@@ -1,55 +1,129 @@
-** When there are user controllable links or hard-coded links to
-third-party domains that specify target="_blank" or target="_window", the a tag should have the rel="noreferrer noopener" attribute specified.
-Allowing users to input markdown is a common culprit, a custom link renderer should be used
-* SSRF - https://www.owasp.org/index.php/Server_Side_Request_Forgery
-All network requests made from the {kib} server should use an explicit configuration or white-list specified in the kibana.yml
\ No newline at end of file
+When writing code for {kib}, be sure to follow these best practices to avoid common vulnerabilities. Refer to the included Open Web
+Application Security Project (OWASP) references to learn more about these types of attacks.
+
+=== Cross-site Scripting (XSS) ===
+
+https://owasp.org/www-community/attacks/xss[_OWASP reference for XSS_]
+
+XSS is a class of attacks where malicious scripts are injected into vulnerable websites. {kib} defends against this primarily by widespread
+use of the React framework to safely encode data that is rendered to pages, widespread use of the EUI framework to
+https://elastic.github.io/eui/#/navigation/link#link-validation[automatically sanitize links], and a restrictive `Content-Security-Policy`
+header.
+
+*Best practices:*
+
+* Check for usage of dangerous functions or assignments that can result in unescaped user input in the browser DOM. Avoid
+using these at all costs:
+** *React:* https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml[`dangerouslySetInnerHtml`].
+** *Browser DOM:* `Element.innerHTML` and `Element.outerHTML`.
+* If using any of the aforementioned unsafe functions/assignments is absolutely necessary, follow
+https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#xss-prevention-rules[these XSS prevention
+rules] to ensure that user input is not inserted into unsafe locations and to make sure that it is escaped properly.
+* Use EUI components to build your UI whenever possible, particularly when rendering `href` links. Otherwise, sanitize user input before
+rendering links to ensure that they do not use the `javascript:` protocol.
+
+=== Cross-Site Request Forgery (CSRF/XSRF) ===
+
+https://owasp.org/www-community/attacks/csrf[_OWASP reference for CSRF_]
+
+CSRF is a class of attacks where a user is forced to execute an action on a vulnerable website that they're logged into, usually without
+their knowledge. {kib} defends against this by requiring
+https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#use-of-custom-request-headers[custom
+request headers] for API endpoints. For more information, see <<api-request-headers, API Request Headers>>.
+
+*Best practices:*
+
+* Ensure all HTTP routes are registered via the <<http-service, {kib} HTTP service>> to take advantage of the aforementioned custom request
```

Comment:
__note:__ probably it's obvious and we don't need to call this out specifically, but I still wanted to note that developers should be very cautious about what they do inside of `GET` API actions (i.e. follow idempotency requirements) as these are not protected with `kbn-xsrf` headers.

Reply:
Good point!